implementation of multivariate resultants.

Nikoleta-v3 opened the pull request:
#### Brief description
Adds a new file `multivariate_resultants` that contains two classes.
These classes are implementations of the following multivariate resultants:

- Dixons
- Macaulay

They are a natural follow up from the resultants implemented withing the library:
- [subresultants_qq_zz.py](https://github.com/sympy/sympy/blob/9670ff89dd8bff7878c6d636ec4f493b87d092df/sympy/polys/subresultants_qq_zz.py)

Resultants are used to indentify if polynomials have common roots. The multivariate version is for multivariate systems.

#### Other comments
Tests have been implemented and are currently passing. I added relevant literature and  some notebooks with examples in the [example/notebooks/](https://github.com/sympy/sympy/tree/master/examples/notebooks) directory.

#### References to other Issues or PRs
This pull request does not fix an issue, neither it implements an ideal from the 
[Project Ideas Page](https://github.com/sympy/sympy/wiki/Project-General-Ideas). I did post about the idea on the google group: https://groups.google.com/forum/#!topic/sympy/22jHQiEoy3Q

Commit:
implementation of multivariate resultants.

Adds a new file `multivariate_resultants` that contains two classes.
These classes are implementations of the following multivariate resultants:

- Dixons
- Macaulay

They are a natural follow up from the resultants implemented withing the
library:
https://github.com/sympy/sympy/blob/9670ff89dd8bff7878c6d636ec4f493b87d092df/sympy/polys/subresultants_qq_zz.py

Resultants are used to indentify if polynomials have common roots.
The multivariate version is for multivariate systems.

Tests have been implemented and are currently passing.
I added relevant literature and  some notebooks with examples in the
`example/notebooks/` directory.

normalhuman commented on sympy/polys/tests/test_multivariate_resultants.py:
```diff
@@ -0,0 +1,195 @@
+c, d = symbols("a, b")
+x, y = symbols("x, y")
+
+p = lambdify((x, y), c * x + y)
```
These are not polynomials, they are Python functions.  `p = Poly(c*x + y, (x, y))` would be a polynomial.

normalhuman commented on sympy/polys/tests/test_multivariate_resultants.py:
```diff
@@ -0,0 +1,195 @@
+
+    p = lambdify((x, y), x + y)
+    q = lambdify((x, y), x ** 2 + y **3)
+    h = lambdify((x, y), x ** 2 + y)
```
Again, these should be polynomials, if this is a part of polynomial module.

normalhuman commented on sympy/polys/multivariate_resultants.py:
```diff
@@ -0,0 +1,390 @@
+                                   for c in coefficients])
+
+        keep = [column for column in range(dixon_matrix.shape[-1])
+                if all([elements == 0 for elements in dixon_matrix[:, column]]) is False]
```
I think the meaning would be more clear in the form `if any([element != 0 for element in dixon_matrix[:, column]])`

normalhuman commented on sympy/polys/multivariate_resultants.py:
```diff
@@ -0,0 +1,390 @@
+class MacaulayResultant():
+    """
+    A class for calculating the Macaulay resultant. Note that the coefficients
+    of the polynomials must be given as symbols.
```
Why? What goes wrong with concrete coefficients like `3*x*y`, etc?

Nikoleta-v3 replied:
For calculating the sub matrix for the Macaulay's resultant we need the
coefficients of  x_i ^ {d_i} which are defined as a_{i}. From the Macaulay's matrix
the rows that contain one of the a_{i}s are dropped. 

If we give the coefficient numerical values and a_{i} has the same value with another coefficient
we won't be able to make a distinction between them.

smichr replied:
Could any numerical values, c, be replaced with `Dummy(str(c))` without causing too much trouble? or have coefficients that are numerical already been removed in some way?

normalhuman reviewed (commented):
This is an interesting addition to SymPy, which needs some work to become an organic part of the library. In particular, the use of `lambdify` to create functions that mimic polynomial objects should be replaced by constructing actual polynomial objects. 

Many polynomial methods in SymPy allow expressions as input, if those can be converted to polynomials; they call `poly_from_expr` on their input. A random example is https://github.com/sympy/sympy/blob/master/sympy/polys/polyfuncs.py#L316 
Perhaps this could be done here as well.

Nikoleta-v3 commented:
@normalhuman Thank you very much for the review and your comments. I will address each one of them as soon as possible.

Commit:
Adressing comments on the pr

This includes the following tweaks:

- fxing typos, docstrings and adds path to notebooks
- removes lambdify, now resultants can handle expressions.
- tweaks notebooks with examples
- remove functool dependancy
- other small tweaks as suggested.

Nikoleta-v3 commented:
@normalhuman I believe I addressed everything.

jksuom commented on sympy/polys/multivariate_resultants.py:
```diff
@@ -0,0 +1,389 @@
+
+
+    def get_coefficients_of_alpha(self, polynomial):
+        r"""
```
How does this compare with `polynomial.coeffs()`?

Nikoleta-v3 replied:
It does not, I should be using `polynomial.coeffs()`. Apologises I missed that.

jksuom commented on sympy/polys/multivariate_resultants.py:
```diff
@@ -0,0 +1,389 @@
+        self.n = len(self.variables)
+        self.m = len(self.polynomials)
+
+        self.dummy_variables = self.get_dummy_variables()
```
Would it be possible to set `dummy_variables` without calling `get_dummy_variables`. It seems that, once it has been set, there is no need for `get_dummy_variables`.

jksuom commented on sympy/polys/multivariate_resultants.py:
```diff
@@ -0,0 +1,389 @@
+        self.m = len(self.polynomials)
+
+        self.dummy_variables = self.get_dummy_variables()
+        self.max_degrees = self.get_max_degrees()
```
Maybe `max_degrees` could be set without `get_max_degrees`.

Nikoleta-v3 replied:
I will remove both methods and just set the variables.

Commit:
removes methods from Dixon's class

jksuom commented on sympy/polys/multivariate_resultants.py:
```diff
@@ -0,0 +1,357 @@
+        self.n = len(self.variables)
+        self.m = len(self.polynomials)
+
+        a = IndexedBase("alpha")
```
It is also possible to define ordinary symbols that print nicely:
```
>>> pprint(symbols('alpha:5'))
(α₀, α₁, α₂, α₃, α₄)
```

Nikoleta-v3 replied:
I hadn't realise that you could define a range of symbols the way you did, thus I used the `IndexedBase` class. Do you think that it makes a big difference? Should I replace the `IndexedBase` with the `symbols` class?

jksuom replied:
I don't think it makes any difference in the operation. There may be some difference in the output.

BTW, the notebook names contain spaces. Those should be avoided. The names could be simply something like "Sylvester's_resultant.ipynb".

Commit:
update example notebooks after removing methods

A few methods have been removed from Dixon's.
Tweaked notebooks to not break.

Commit:
renaming notebooks

jksuom commented on sympy/polys/multivariate_resultants.py:
```diff
@@ -0,0 +1,357 @@
+This module contains functions for two multivariate resultants. These are:
+
+- Dixon's resultant.
+- Maucalay's resultant.
```
Macalay -> Macaulay

jksuom commented on sympy/polys/multivariate_resultants.py:
```diff
@@ -0,0 +1,357 @@
+        self.dummy_variables = [a[i] for i in range(self.n)]
+
+        # A list of the d_max of each variable.
+        self.max_degrees = [poly_from_expr(f, self.variables)[0].degree() for f in self.polynomials]
```
This line could be split (probably after degree()).

jksuom commented on sympy/polys/multivariate_resultants.py:
```diff
@@ -0,0 +1,357 @@
+        iterator = iter(self.dummy_variables)
+
+        for idx in range(self.n):
+            temp[idx] = next(iterator)
```
Could this be `temp[idx] = self.dummy_variables[idx]`?

jksuom commented on sympy/polys/multivariate_resultants.py:
```diff
@@ -0,0 +1,357 @@
+
+        for idx in range(self.n):
+            temp[idx] = next(iterator)
+            substitution = {self.variables[i]: temp[i] for i in range(len(temp))}
```
Could this be something like `substitution = {var: t for var, t in zip(self.variables, temp)}`?

Commit:
fixes typo and cleans code

Addressing the comments on the PR

Nikoleta-v3 commented:
@jksuom and @normalhuman thank you very much for your reviews. Please let me know if there is anything else you would like me to address.

jksuom commented:
There are some minor methods like `get_max_degrees` and `get_polynomial_degree` that are probably already implemented in SymPy. Would it be possible to use those?

Commit:
removew get_max and get_polynoliam methods

asmeurer commented:
Can you change the notebooks to only use ASCII characters in their filenames.

asmeurer commented on examples/notebooks/Bezout's_Dixon's_resultants.ipynb:
```diff
@@ -0,0 +1,970 @@
+   "metadata": {},
+   "source": [
+    "$$\\left|\\begin{matrix}\n",
+    "p(x) & q(x)\\cr\n",
```
This is showing "unknown character" in [GitHub](https://github.com/Nikoleta-v3/sympy/blob/b4ca77bed357e3d3091e9b70fc249b0a05895b6c/examples/notebooks/Bezout's_Dixon's_resultants.ipynb)

asmeurer commented:
Note for reviewers: if you click "view" next to the notebook filename on the PR diff, it will open the notebook in a rendered view.

asmeurer commented on examples/notebooks/Sylvester's_resultants, .ipynb:
```diff
@@ -0,0 +1,714 @@
+    "\n",
+    "then the resultant $R(p,q)$ of $p$ and $q$ is defined as:\n",
+    "\n",
+    "$$R(p,q)=a^n_{0}b^m_{0}\\prod_{i=1}^{m}\\prod_{j=1}^{n}(r_i−s_j)$$"
```
There are also a lot of "unknown characters" in this notebook.

Commit:
renames notebooks

Commit:
fixes unknown characters in notebooks

Nikoleta-v3 commented:
@asmeurer apologises for not realising that there were  "unknown characters". It was rendering correctly locally so I missed it. I fixed all the parts that were breaking and renamed the notebooks.

jksuom commented:
It looks like there is a space in Sylvester resultant notebook's name

Commit:
removes space from `Sylvester_resultant` notebook

jksuom commented on sympy/polys/multivariate_resultants.py:
```diff
@@ -0,0 +1,341 @@
+            The degree_m is calculated as  1 + \sum_1 ^ n (d_i - 1), where
+            d_i is the degree of the i polynomial
+        """
+        return 1 + sum([d - 1 for d in self.degrees])
```
This could be `1 + sum(d - 1 for d in self.degrees)` (generator instead of a list).

Commit:
replaces list with generator and fixes typo

jksuom commented on sympy/polys/multivariate_resultants.py:
```diff
@@ -0,0 +1,341 @@
+        self.n = len(variables)
+
+        # A list of the d_max of each variable.
+        self.degrees = [poly_from_expr(poly, self.variables)[0].degree()
```
Could this be `total_degree(poly, *variables)`?

jksuom commented on sympy/polys/multivariate_resultants.py:
```diff
@@ -0,0 +1,341 @@
+        self.degree_m = self.get_degree_m()
+        self.monomials_size = self.get_size()
+
+    def get_degree_m(self):
```
Is this used other than setting `self.degree_m`?

Nikoleta-v3 replied:
No the method is only used for setting the `self.degree_m`. It is implemented as method because I wanted to use the docstring to explain how `degree_m` is calculated. Also I wanted to be able to test it.

asmeurer replied:
It should probably be called _get_degree_m then, so that it is private.

Nikoleta-v3 replied:
I will do that, thank you!

Commit:
makes use of total_degree method

jksuom commented on sympy/polys/multivariate_resultants.py:
```diff
@@ -0,0 +1,338 @@
+        monomials: list
+            A list of monomials of a certain degree.
+        """
+        monomials = list(itermonomials(self.variables, degree) -
```
This could be implemented more efficiently by direct use of `combinations_with_replacement` as in `itermonomials` but without adding 1 to the list of variables.

jksuom commented on sympy/polys/multivariate_resultants.py:
```diff
@@ -0,0 +1,338 @@
+        monomials = list(itermonomials(self.variables, degree) -
+                         itermonomials(self.variables, degree - 1))
+
+        return sorted(monomials, key=monomial_key('lex', self.variables))[::-1]
```
With `reverse=True`, the reversed copy is created by `sorted`.

Commit:
makes get_degree_m a private method

Commit:
sorts monomials using reverse

Commit:
uses combinations_with_replacement

Instead of retrieving two lists of monomials
and substracting them to the monomials of centrain degree,
uses the combinations_with_replacement function.

jksuom commented on sympy/polys/multivariate_resultants.py:
```diff
@@ -0,0 +1,339 @@
+                poss_rows = self.get_monomials_of_certain_degree(self.degree_m - self.degrees[i])
+                for div in divisible:
+                    for p in poss_rows:
+                        if fraction((p / div).expand())[1] == 1:
```
To see that `p` is divisible by `div`, we can check the remainder: `rem(p, div) == 0`.

Commit:
uses function rem

Abdullahjavednesar commented:
Ping @jksuom, can you review this, thanks!

jksuom commented:
I have not been able to look into the references. Would it be possible to add some that are easily available? There are also several long lines that should be split (preferably at most 72 characters). Otherwise I think this would be ready to be merged.

Nikoleta-v3 commented:
Thank you @jksuom . I will address both your comments asap.

Commit:
pep8 tweaks

Commit:
adds links to references.

The links are to open copies.

jksuom commented on sympy/polys/multivariate_resultants.py:
```diff
@@ -110,44 +114,50 @@ def get_dixon_polynomial(self):
-        list_of_products = [self.variables[i] ** ((i + 1) * self.max_degrees[i] - 1)
-                            for i in range(self.n)]
+        list_of_products = [self.variables[i] ** ((i + 1) * \
+                            self.max_degrees[i] - 1) for i in range(self.n)]
```
There is no need for backslash continuation as everything is surrounded by brackets.

PEP 8:

> The preferred way of wrapping long lines is by using Python's implied line continuation inside parentheses, brackets and braces. Long lines can be broken over multiple lines by wrapping expressions in parentheses. These should be used in preference to using a backslash for line continuation.

jksuom commented on sympy/polys/multivariate_resultants.py:
```diff
@@ -255,14 +270,19 @@ def get_row_coefficients(self):
             else:
-                divisible.append(self.variables[i - 1] ** self.degrees[i - 1])
-                poss_rows = self.get_monomials_of_certain_degree(self.degree_m - self.degrees[i])
+                divisible.append(self.variables[i - 1] ** \
```
Here as well.

Commit:
removes blackslash

Abdullahjavednesar commented:
Seems all good, hope this is ready now @jksuom.

jksuom commented:
Yes, I think this is ready to be merged. Thank you, @Nikoleta-v3

Abdullahjavednesar commented:
Great work, thanks @Nikoleta-v3, @jksuom

Nikoleta-v3 commented:
Thank you everyone.
